### PR TITLE
ICU-20488 addition mutex static constructor fixes

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20488
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by aheninger** [comment]
Hmm, if were hitting an environment where the existing code doesn't work, then I guess we need to change it.

I had hoped to avoid allocating memory that never is deleted; while it sneaks by the existing set of checks, it seems likely to eventually trip us up with improved or stronger leak checkers. 

Your proposed fix is, I think, otherwise OK.

This PR is opened against master, not maint/maint-64. It's very late for any 64 changes; we need to check with Markus on how to proceed. Don't merge yet.

**Comment by markusicu** [comment]
Too late for ICU 64.1. We just have to stop somewhere and move on with the release tasks.

For this issue, the concern seems to be that the mutex destructors do something, presumably releasing an OS mutex handle. If some other destructor was using a mutex that had been destructed earlier, it could crash. However, these are initializer and cache mutexes -- is there any real danger here? The static-pointer version avoids this altogether by never calling the destructor.

I share Andy's concern about leak checkers. We check that there are no leaks by running both test suites and making sure that there is no leaked heap memory at the end. Will valgrind complain about local static pointers?

**Comment by aheninger** [comment]
> Will valgrind complain about local static pointers?

Almost certainly yes, although I haven't run it yet.
Address sanitizer does not flag it - it allows leaks that are accessible via globals, and only flags unreachable stuff.

**Comment by aheninger** [comment]
I think a better thing would be to get rid of the UMutex class completely, and use std::mutex directly in ICU code.
UMutex's purpose was to hide the OS dependent mutex implementations - pthread, Windows critical sections, or whatever. It's no longer needed.

The  std::mutex docs say this:

> Because the default constructor is constexpr, static mutexes are initialized as part of static non-local initialization, before any dynamic non-local initialization begins. This makes it safe to lock a mutex in a constructor of any static object.

Apple has problems with this, but Linux is good, and V8 appears to be good because this change didn't find the places where ICU does directly uses a static std::mutex.

**Comment by pedberg-icu** [comment]
It looks like the changes here omitted the parallel change in copyInvariantFormatter in io/locbund.cp; shouldn't that also have been changed to use new UMutex() etc.?

File: icu4c/source/common/umutex.cpp [hunk 1]
```diff
@@ -43,8 +43,8 @@ U_NAMESPACE_BEGIN
 
 // The ICU global mutex. Used when ICU implementation code passes NULL for the mutex pointer.
 static UMutex *globalMutex() {
-    static UMutex m = U_MUTEX_INITIALIZER;
-    return &m;
+    static UMutex *m = new UMutex();
```
Comment (axelandrejs) [inline icu4c/source/common/umutex.cpp]: >static UMutex *m = new UMutex(); [](start = 4, length = 32)

This leaks the mutex if the binary containing ICU is ever unloaded.

**Comment by FrankYFTang** [comment]
obsoleted

**Pull request closed without merging**